File: cube_petit_gazebo/src/docking_station_simulation.cpp
#include "../include/docking_station_simulation.hpp"

/////////////////////////////
/*  コンストラクタ         */
/////////////////////////////

// Docking_Station_Simulation::Docking_Station_Simulation(ros::NodeHandle nh):
//   battery_charge_simulation(nh)

Docking_Station_Simulation::Docking_Station_Simulation(ros::NodeHandle nh){
  ROS_INFO("Docking_Station_Simulation::Docking_Station_Simulation ->start");
  // get rosparam (debug)
  publish_debug_topics_flag = true;
  nh.getParam("publish_debug_topics", publish_debug_topics_flag);
  // get rosparam (link_name)
  robot_link_name = "base";
  station_link_name = "station";
  nh.getParam("robot_link_name", robot_link_name);
  nh.getParam("station_link_name", station_link_name);
  // get rosparam (docking param)
  robot_connector_offset = 0.18;
  station_connector_offset = 0.22;
  docking_area_radious = 0.03;
  docking_area_angle = 0.5;
  voltage = 30.0;
  nh.getParam("robot_connector_offset", robot_connector_offset);
  nh.getParam("station_connector_offset", station_connector_offset);
  nh.getParam("docking_area_radious", docking_area_radious);
  nh.getParam("docking_area_angle", docking_area_angle);
  nh.getParam("voltage", voltage);
  //publisher
  output_volatage_pub = nh.advertise<std_msgs::Float32>("output_voltage", 10);
  output_current_pub = nh.advertise<std_msgs::Float32>("output_current", 10);
  if(publish_debug_topics_flag){
    robot_connector_pose_pub = nh.advertise<geometry_msgs::PoseStamped>("debug/robot_pose",5);
    station_connector_pose_pub = nh.advertise<geometry_msgs::PoseStamped>("debug/station_pose",5);
    target_pose_pub = nh.advertise<geometry_msgs::PoseStamped>("debug/target_pose",5);
    docking_flag_pub = nh.advertise<std_msgs::Bool>("debug/docking_flag", 5);
  }
  gazebo_robot_odom = *(ros::topic::waitForMessage<nav_msgs::Odometry>("robot_odom", nh));
  gazebo_station_odom = *(ros::topic::waitForMessage<nav_msgs::Odometry>("station_odom", nh));  
  //subscriber
  //[TODO]
  robot_odom_sub = nh.subscribe("robot_odom", 10, &Docking_Station_Simulation::robotPoseSubscriberCallback, this);
  station_odom_sub  = nh.subscribe("station_odom", 10, &Docking_Station_Simulation::stationPoseSubscriberCallback, this);
  ROS_INFO("Docking_Station_Simulation::Docking_Station_Simulation ->done");

}

Docking_Station_Simulation::~Docking_Station_Simulation() {

}


void Docking_Station_Simulation::publishDataLoop(){
  bool docking_flag = validateDocking(getConnectorDevitation());
  std_msgs::Float32 voltage_message;
  std_msgs::Float32 current_message;
  if(docking_flag){
    voltage_message.data = voltage;
    current_message.data = 5.0;
  }else{
    voltage_message.data = 0.0;
    current_message.data = -3.0;
  }
  output_current_pub.publish(current_message);
  output_volatage_pub.publish(voltage_message);
}

geometry_msgs::PoseStamped Docking_Station_Simulation::getConnectorDevitation(){
  geometry_msgs::Quaternion quaternion_world2robot;
  quaternion_world2robot.x = gazebo_robot_odom.pose.pose.orientation.x;
  quaternion_world2robot.y = gazebo_robot_odom.pose.pose.orientation.y;
  quaternion_world2robot.z = gazebo_robot_odom.pose.pose.orientation.z;
  quaternion_world2robot.w = gazebo_robot_odom.pose.pose.orientation.w;

  float yaw_world2robot = quaternion2yaw(quaternion_world2robot);  
  Eigen::Matrix2f  rotation_matrix_world2robot_vec2;
  Eigen::Matrix2f  rotation_matrix_world2robot_vec2_inv;
  rotation_matrix_world2robot_vec2 = yaw2matrix2(yaw_world2robot);
  rotation_matrix_world2robot_vec2_inv = yaw2matrix2(-yaw_world2robot);

  Eigen::Vector2f world2robot_vec2;
  Eigen::Vector2f world2station_vec2;
  world2robot_vec2(0) = gazebo_robot_odom.pose.pose.position.x;
  world2robot_vec2(1) = gazebo_robot_odom.pose.pose.position.y;
  world2station_vec2(0) = gazebo_station_odom.pose.pose.position.x;
  world2station_vec2(1) = gazebo_station_odom.pose.pose.position.y;
  
  Eigen::Vector2f robot2station_vec2;
  robot2station_vec2(0) = world2station_vec2(0) - world2robot_vec2(0);
  robot2station_vec2(1) = world2station_vec2(1) - world2robot_vec2(1);

  Eigen::Vector2f rotated_robot2station2;
  rotated_robot2station2 = rotation_matrix_world2robot_vec2_inv * robot2station_vec2;

  geometry_msgs::Quaternion quaternion_world2station;
  quaternion_world2station.x = gazebo_station_odom.pose.pose.orientation.x;
  quaternion_world2station.y = gazebo_station_odom.pose.pose.orientation.y;
  quaternion_world2station.z = gazebo_station_odom.pose.pose.orientation.z;
  quaternion_world2station.w = gazebo_station_odom.pose.pose.orientation.w;

  float yaw_world2station = quaternion2yaw(quaternion_world2station);  
  Eigen::Matrix2f rotation_matrix_world2station2;
  rotation_matrix_world2station2 = yaw2matrix2(yaw_world2station);

  float yaw_robot2station = yaw_world2station - yaw_world2robot;

  //robotのPose
  geometry_msgs::PoseStamped robot2robot;

  if(publish_debug_topics_flag){
    geometry_msgs::Quaternion quaternion_robot2station = yaw2quaternion(yaw_robot2station);

    //stationのPose
    geometry_msgs::PoseStamped robot2station;
    robot2station.header = gazebo_station_odom.header;
    robot2station.pose = gazebo_station_odom.pose.pose;
    //[謎]
    robot2station.header.frame_id = gazebo_robot_odom.child_frame_id;
    robot2station.pose.position.x = rotated_robot2station2(0);
    robot2station.pose.position.y = rotated_robot2station2(1);
    robot2station.pose.position.z = 0;
    robot2station.pose.orientation.x = quaternion_robot2station.x;
    robot2station.pose.orientation.y = quaternion_robot2station.y;
    robot2station.pose.orientation.z = quaternion_robot2station.z;
    robot2station.pose.orientation.w = quaternion_robot2station.w;

    robot2robot.header = gazebo_robot_odom.header;
    robot2robot.pose = gazebo_robot_odom.pose.pose;
    robot2robot.header.frame_id = gazebo_robot_odom.child_frame_id;
    robot2robot.pose.position.x = 0;
    robot2robot.pose.position.y = 0;
    robot2robot.pose.position.z = 0;
    robot2robot.pose.orientation.x = 0;
    robot2robot.pose.orientation.y = 0;
    robot2robot.pose.orientation.z = 0;
    robot2robot.pose.orientation.w = 1;

    robot_connector_pose_pub.publish(robot2robot);
    station_connector_pose_pub.publish(robot2station);

  }

  Eigen::Vector2f robot2connector_vec;
  robot2connector_vec(0) = robot_connector_offset;
  robot2connector_vec(1) = 0;

  Eigen::Vector2f world2robot_con_vec;
  world2robot_con_vec = world2robot_vec2 + rotation_matrix_world2robot_vec2 * robot2connector_vec;

  Eigen::Vector2f station2connector_vec;
  station2connector_vec(0) = station_connector_offset;
  station2connector_vec(1) = 0;

  Eigen::Vector2f world2station_con_vec;
  world2station_con_vec = world2station_vec2 + rotation_matrix_world2station2 * station2connector_vec;

  Eigen::Vector2f robot2station_con_vec;
  robot2station_con_vec = world2station_con_vec - world2robot_con_vec;

  Eigen::Vector2f rotated_robot2station_con_vec;
  rotated_robot2station_con_vec = rotation_matrix_world2robot_vec2_inv * robot2station_con_vec;
  // ROS_INFO("vec x: %f, y: %f",rotated_robot2station_con_vec(0),rotated_robot2station_con_vec(1));

  geometry_msgs::Quaternion quaternion_robot2target;
  quaternion_robot2target = yaw2quaternion(yaw_robot2station + M_PI);

  //targetのPose
  geometry_msgs::PoseStamped robot2station_con_pose_stamped;
  robot2robot.header = gazebo_robot_odom.header;
  robot2robot.pose = gazebo_robot_odom.pose.pose;
  robot2station_con_pose_stamped.header.frame_id = gazebo_robot_odom.child_frame_id;
  robot2station_con_pose_stamped.pose.position.x = rotated_robot2station_con_vec(0);
  robot2station_con_pose_stamped.pose.position.y = rotated_robot2station_con_vec(1);
  robot2station_con_pose_stamped.pose.position.z = 0;
  robot2station_con_pose_stamped.pose.orientation.x = quaternion_robot2target.x;
  robot2station_con_pose_stamped.pose.orientation.y = quaternion_robot2target.y;
  robot2station_con_pose_stamped.pose.orientation.z = quaternion_robot2target.z;
  robot2station_con_pose_stamped.pose.orientation.w = quaternion_robot2target.w;
  if(publish_debug_topics_flag){
    target_pose_pub.publish(robot2station_con_pose_stamped);
  }
  return robot2station_con_pose_stamped;
}

bool Docking_Station_Simulation::validateDocking(geometry_msgs::PoseStamped robot2station_connectors){
  float translation_diff = sqrt(pow(robot2station_connectors.pose.position.x,2) + pow(robot2station_connectors.pose.position.y, 2));
  
  geometry_msgs::Quaternion anguler_diff_quaternion;
  anguler_diff_quaternion.x = robot2station_connectors.pose.orientation.x;
  anguler_diff_quaternion.y = robot2station_connectors.pose.orientation.y;
  anguler_diff_quaternion.z = robot2station_connectors.pose.orientation.z;
  anguler_diff_quaternion.w = robot2station_connectors.pose.orientation.w;

  //absにintが返されるので違う気がする
  float ang_diff_temp = quaternion2yaw(anguler_diff_quaternion);
  if(ang_diff_temp < 0.0){
    ang_diff_temp = ang_diff_temp * (-1.0);
  }
  float anguler_diff = ang_diff_temp;
  bool docking_flag = false;

  // ROS_INFO("yaw: %f", quaternion2yaw(anguler_diff_quaternion));

  // ROS_INFO("trans: %f ( < %f ), ang: %f (  > %f )", translation_diff, docking_area_radious, anguler_diff, docking_area_angle);

  if( translation_diff < docking_area_radious && anguler_diff < docking_area_angle){
    docking_flag = true;
  }
  if(publish_debug_topics_flag){
    std_msgs::Bool docking_flag_msgs;
    docking_flag_msgs.data = docking_flag;
    //publish
    docking_flag_pub.publish(docking_flag_msgs);
  }

  return docking_flag;
}

float Docking_Station_Simulation::quaternion2yaw(geometry_msgs::Quaternion quoternion){
  float x = quoternion.x;
  float y = quoternion.y;
  float z = quoternion.z;
  float w = quoternion.w;
  float yaw_ret = atan2(2.0 * (w * z + x * y), w * w + x * x - y * y - z * z);
  return yaw_ret;
}

void Docking_Station_Simulation::yaw2matrix( std::vector<std::vector<float> > &rot, float yaw){
  float cos_f = cos(yaw);
  float sin_f = sin(yaw);
  float rot_ret[2][2] = {{cos_f, -sin_f}, {sin_f, cos_f}};
  rot[0][0] = cos_f;
  rot[0][1] = -sin_f;
  rot[1][0] = sin_f;
  rot[1][1] = cos_f;
}

Eigen::Matrix2f Docking_Station_Simulation::yaw2matrix2(float yaw){
  Eigen::Matrix2f rot;
  float cos_f = cos(yaw);
  float sin_f = sin(yaw);
  float rot_ret[2][2] = {{cos_f, -sin_f}, {sin_f, cos_f}};
  rot(0,0) = cos_f;
  rot(0,1) = -sin_f;
  rot(1,0) = sin_f;
  rot(1,1) = cos_f;
  return rot;
}

geometry_msgs::Quaternion Docking_Station_Simulation::yaw2quaternion(float yaw){
  geometry_msgs::Quaternion ret_quaternion;
  ret_quaternion.x = 0;
  ret_quaternion.y = 0;
  ret_quaternion.z = cos(yaw/2.0);
  ret_quaternion.w = sin(yaw/2.0);
  return ret_quaternion;
}

void Docking_Station_Simulation::robotPoseSubscriberCallback(const nav_msgs::Odometry::ConstPtr &odom_data){
  gazebo_robot_odom = *odom_data;
}

void Docking_Station_Simulation::stationPoseSubscriberCallback(const nav_msgs::Odometry::ConstPtr &odom_data){
  gazebo_station_odom = *odom_data;
}

int main(int argc, char **argv){
  ros::init(argc, argv, "docking_statoion_simulation");

  ros::NodeHandle nh("~");

  Docking_Station_Simulation docking_station_simulation(nh);

  ros::Rate loop_rate(20);

  while(ros::ok()){
    docking_station_simulation.publishDataLoop();
    ros::spinOnce();
    loop_rate.sleep();
  }
  return 0;
}

File: cube_petit_gazebo/scripts/docking_station_simulation.py
#!/usr/bin/env python

from math import pi, sqrt
import rospy
from std_msgs.msg import Float32, Bool
from nav_msgs.msg import Odometry
from geometry_msgs.msg import PoseStamped
import numpy as np


class DockingStationSimulation:

    def __init__(self):
        rospy.logdebug('[DockingStationSimulation] Start Initialization')
        self.rate = rospy.Rate(20)
        # Get rosparam
        self.publish_debug_topics = rospy.get_param(
            '~publish_debug_topics', True)
        # Link names
        # Robot link name published by Gazebo plugin
        self.robot_link_name = rospy.get_param(
            '~robot_link_name', 'base')
        # Charging-station link name published by Gazebo plugin
        self.station_link_name = rospy.get_param(
            '~station_link_name', 'station')
        # Docking parameters
        # Distance from robot origin to the charging connector in x-axis
        # direction [m]
        self.robot_connector_offfset = rospy.get_param(
            '~robot_connector_offfset', 0.18)
        # Distance from charging-station origin to the charging connector in
        # the x-axis direction [m]
        self.station_connector_offset = rospy.get_param(
            '~station_connector_offset', 0.22)
        # Radius of the area which publish the charging voltage [m]
        self.docking_area_radious = rospy.get_param(
            '~docking_area_radious', 0.03)
        # Range of angles to publish charging voltage (+-
        # 'publih_voltage_area_angle' [m])
        self.docking_area_angle = rospy.get_param('~docking_area_angle', 0.5)
        # Charging Voltage
        self.voltage = rospy.get_param('~voltage', 30.0)
        # Initialize publishers
        self.output_voltage_pub = rospy.Publisher(
            '~output_voltage', Float32, queue_size=5)
        self.output_current_pub = rospy.Publisher(
            '~output_current', Float32, queue_size=5)
        if self.publish_debug_topics:
            self.robot_connector_pose_pub = rospy.Publisher(
                '~debug/robot_pose', PoseStamped, queue_size=5)
            self.station_connector_pose_pub = rospy.Publisher(
                '~debug/station_pose', PoseStamped, queue_size=5)
            self.targeet_pose_pub = rospy.Publisher(
                '~debug/target_pose', PoseStamped, queue_size=5)
            self.docking_flag_pub = rospy.Publisher(
                '~debug/docking_flag', Bool, queue_size=5)
        # Wait for gazebo plugin topics
        rospy.logdebug('[DockingStationSimulation] Waiting for Gazebo topics')
        self.gazebo_robot_odom = rospy.wait_for_message(
            '~robot_odom', Odometry)
        self.gazebo_station_odom = rospy.wait_for_message(
            '~station_odom', Odometry)
        # Initialize subscribers
        rospy.Subscriber('~robot_odom', Odometry,
                         self.robot_pose_subscriber_callback)
        rospy.Subscriber('~station_odom', Odometry,
                         self.station_pose_subscriber_callback)
        rospy.logdebug('[DockingStationSimulation] Finish Initialization')

    def publish_data_loop(self):
        self.rate.sleep()
        docking_flag = self.validate_docking(
            self.get_connector_deviation())
        voltage_message = Float32()
        current_message = Float32()
        if docking_flag:
            voltage_message.data = self.voltage
            current_message.data = 5.0
        else:
            voltage_message.data = 0.0
            current_message.data = -3.0
        self.output_voltage_pub.publish(voltage_message)
        self.output_current_pub.publish(current_message)

    def get_connector_deviation(self):
        # Calculate position (robot to station) on x-y plane
        quaternion_world2robot = np.array([
            self.gazebo_robot_odom.pose.pose.orientation.x,
            self.gazebo_robot_odom.pose.pose.orientation.y,
            self.gazebo_robot_odom.pose.pose.orientation.z,
            self.gazebo_robot_odom.pose.pose.orientation.w])
        yaw_world2robot = self.quaternion2yaw(quaternion_world2robot)
        rotation_matrix_world2robot = self.yaw2matrix(yaw_world2robot)
        rotation_matrix_world2robot_inv = self.yaw2matrix(- yaw_world2robot)
        vector_world2robot = np.array([
            self.gazebo_robot_odom.pose.pose.position.x,
            self.gazebo_robot_odom.pose.pose.position.y])
        vector_world2station = np.array([
            self.gazebo_station_odom.pose.pose.position.x,
            self.gazebo_station_odom.pose.pose.position.y])
        vector_robot2station = vector_world2station - vector_world2robot
        rotated_robot2station = np.dot(
            rotation_matrix_world2robot_inv, vector_robot2station)
        # Calculate orientation (robot to station) on x-y plane
        quaternion_world2station = np.array([
            self.gazebo_station_odom.pose.pose.orientation.x,
            self.gazebo_station_odom.pose.pose.orientation.y,
            self.gazebo_station_odom.pose.pose.orientation.z,
            self.gazebo_station_odom.pose.pose.orientation.w])
        yaw_world2station = self.quaternion2yaw(quaternion_world2station)
        rotation_matrix_world2station = self.yaw2matrix(yaw_world2station)
        yaw_robot2station = yaw_world2station - yaw_world2robot
        # rotation_matrix_robot2station = self.yaw2matrix(yaw_robot2station)

        if self.publish_debug_topics:
            quaternion_robot2station = self.yaw2quaternion(yaw_robot2station)
            # Set ROS topic (PoseStamped)
            robot2station = PoseStamped()
            robot2station.header = self.gazebo_station_odom.header
            robot2station.pose = self.gazebo_station_odom.pose.pose
            robot2station.header.frame_id = self.gazebo_robot_odom.child_frame_id
            robot2station.pose.position.x = rotated_robot2station[0]
            robot2station.pose.position.y = rotated_robot2station[1]
            robot2station.pose.position.z = 0
            robot2station.pose.orientation.x = quaternion_robot2station[0]
            robot2station.pose.orientation.y = quaternion_robot2station[1]
            robot2station.pose.orientation.z = quaternion_robot2station[2]
            robot2station.pose.orientation.w = quaternion_robot2station[3]
            robot2robot = PoseStamped()
            robot2robot.header = self.gazebo_robot_odom.header
            robot2robot.pose = self.gazebo_robot_odom.pose.pose
            robot2robot.header.frame_id = self.gazebo_robot_odom.child_frame_id
            robot2robot.pose.position.x = 0
            robot2robot.pose.position.y = 0
            robot2robot.pose.position.z = 0
            robot2robot.pose.orientation.x = 0
            robot2robot.pose.orientation.y = 0
            robot2robot.pose.orientation.z = 0
            robot2robot.pose.orientation.w = 1
            self.robot_connector_pose_pub.publish(robot2robot)
            self.station_connector_pose_pub.publish(robot2station)
        # Calculate connector offsets
        vector_robot2connector = np.array(
            [self.robot_connector_offfset, 0])
        vector_world2robot_con = vector_world2robot + \
            np.dot(rotation_matrix_world2robot, vector_robot2connector)
        vector_station2connector = np.array(
            [self.station_connector_offset, 0])
        vector_world2station_con = vector_world2station + \
            np.dot(rotation_matrix_world2station, vector_station2connector)
        vector_robot2station_con = vector_world2station_con - vector_world2robot_con
        rotated_robot2station_con = np.dot(
            rotation_matrix_world2robot_inv, vector_robot2station_con)
        # Calculate target angle (opposite side of the charging station)
        # rotation_matrix_robot2target = self.yaw2matrix(yaw_robot2station + pi)
        quaternion_robot2target = self.yaw2quaternion(yaw_robot2station + pi)
        # Set ROS topic (PoseStamped)
        robot2station_con = PoseStamped()
        robot2robot.header = self.gazebo_robot_odom.header
        robot2robot.pose = self.gazebo_robot_odom.pose.pose
        robot2station_con.header.frame_id = self.gazebo_robot_odom.child_frame_id
        robot2station_con.pose.position.x = rotated_robot2station_con[0]
        robot2station_con.pose.position.y = rotated_robot2station_con[1]
        robot2station_con.pose.position.z = 0
        robot2station_con.pose.orientation.x = quaternion_robot2target[0]
        robot2station_con.pose.orientation.y = quaternion_robot2target[1]
        robot2station_con.pose.orientation.z = quaternion_robot2target[2]
        robot2station_con.pose.orientation.w = quaternion_robot2target[3]
        if self.publish_debug_topics:
            self.targeet_pose_pub.publish(robot2station_con)
        return robot2station_con

    def validate_docking(self, robot2station_connectors):
        translation_diff = sqrt(
            robot2station_connectors.pose.position.x ** 2
            + robot2station_connectors.pose.position.y ** 2)
        anguler_diff_quaternion = np.array([
            robot2station_connectors.pose.orientation.x,
            robot2station_connectors.pose.orientation.y,
            robot2station_connectors.pose.orientation.z,
            robot2station_connectors.pose.orientation.w])
        anguler_diff = abs(self.quaternion2yaw(anguler_diff_quaternion))
        docking_flag = False
        if translation_diff < self.docking_area_radious and anguler_diff > self.docking_area_angle:
            docking_flag = True
        if self.publish_debug_topics:
            docking_flag_msg = Bool()
            docking_flag_msg.data = docking_flag
            self.docking_flag_pub.publish(docking_flag_msg)
        return docking_flag

    def quaternion2yaw(self, quaternion=[0, 0, 0, 1]):
        x = quaternion[0]
        y = quaternion[1]
        z = quaternion[2]
        w = quaternion[3]
        yaw = np.arctan2(2.0 * (w * z + x * y), w * w + x * x - y * y - z * z)
        return yaw

    def yaw2matrix(self, yaw=0):
        cos = np.cos(yaw)
        sin = np.sin(yaw)
        rot = np.array([[cos, -sin], [sin, cos]])
        return rot

    def yaw2quaternion(self, yaw=0):
        cos = np.cos(yaw / 2.0)
        sin = np.sin(yaw / 2.0)
        quaternion = [0, 0, sin, cos]
        return quaternion

    def robot_pose_subscriber_callback(self, data):
        rospy.logdebug_throttle(
            1,
            '[DockingStationSimulation] robot_pose_subscriber_callback called (show every 1 second)')
        self.gazebo_robot_odom = data

    def station_pose_subscriber_callback(self, data):
        rospy.logdebug_throttle(
            1,
            '[DockingStationSimulation] station_pose_subscriber_callback called (show every 1 second)')
        self.gazebo_station_odom = data


if __name__ == '__main__':
    # Initialize node
    rospy.init_node('charging_station_simulation', log_level=rospy.INFO)
    charging_station_simulation = DockingStationSimulation()
    while not rospy.is_shutdown():
        charging_station_simulation.publish_data_loop()
